src/metrics.rs

```diff
@@ -339,19 +368,61 @@ impl MetricsLayer {
         MetricsLayer {
             meter,
             instruments: Default::default(),
+            callsites: Callsites::new(),
         }
     }
+
+    fn has_metrics_fields(&self, field_set: &FieldSet) -> bool {
+        field_set.iter().any(|field| {
+            let name = field.name();
+            name.starts_with(METRIC_PREFIX_COUNTER)
+                || name.starts_with(METRIC_PREFIX_MONOTONIC_COUNTER)
+                || name.starts_with(METRIC_PREFIX_HISTOGRAM)
+        })
+    }
 }
 
 impl<S> Layer<S> for MetricsLayer
 where
     S: Subscriber + for<'span> LookupSpan<'span>,
 {
     fn on_event(&self, event: &tracing::Event<'_>, _ctx: Context<'_, S>) {
```

Comment:
To avoid allocating each time `on_event()` is called, we determine if the event has a metrics field.
However, this caused RwLock::read() to be called for each `on_event()`.

I am concerned that an implementation that gets read lock for every on_event may not be acceptable.

Reply:
Yeah we may want to see if there is another way, or have this be an opt-in for non-performance sensitive cases

Reply:
I will try [per layer filtering](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/#per-layer-filtering)